docs/source/reference/build_config.rst

```diff
@@ -0,0 +1,357 @@
+    give a quick overview of how Jinja2 is used in conjunction with BentoML. For
+    any reference on Jinja2 please refers to their `Templates Design Documentation <https://jinja.palletsprojects.com/en/3.1.x/templates/>`_.
+
+.. tip::
+
+    One huge advantage of this feature is that it allows developers and library,
+    such as `🚀 bentoctl <https://github.com/bentoml/bentoctl>`_ to extend a 🍱.
+
+The Dockerfile generation process is as follows:
+    - Retrieve options provided via :code:`bentofile.yaml` :code:`docker` section.
+    - Determine a given :class:`DistroSpec` for the coresponding distros. This
+      includes supported python version, cuda version, and architecture type.
+    - From the given spec, alongside with options from :class:`BentoBuildConfig`,
+      generate a Dockerfile from a set of Jinja templates.
+
+One of the powerful features Jinja offers is its `template inheritance <https://jinja.palletsprojects.com/en/3.1.x/templates/#template-inheritance>`_.
+This allows BentoML to enable users to fully customize how to structure a 🍱's Dockerfile.
+
+.. note::
+
+    To use a custom Dockerfile template, users have to provide a file with a format
+    that follows the Jinja2 template syntax. The template file should have
+    extensions of :code:`.j2`, :code:`.template`, :code:`.jinja`.
+
+In addition to Jinja, BentoML also uses `Docker Buildx <https://docs.docker.com/desktop/multi-arch/>`_, which enables users to build Bentos that support
+multiple architectures. Under the hood the generated Dockefile leverage
+`buildkit <https://github.com/moby/buildkit>`_, which enables some advanced `dockerfile features <https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#dockerfile-frontend-syntaxes>`_.
+
+.. note::
+
+   Currently the only buildkit frontend that we are supporting is
+   :code:`docker/dockerfile:1.4-labs`. Please contact us if you need support for
+   different build frontend.
+
+
+Writing custom Dockerfile template
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To construct a custom :code:`Dockerfile` template, users have to provide the block below at the beginning of the Dockerfile template :code:`Dockerfile.template`:
+
+.. code-block:: jinja
+
+    {% extends bento__dockerfile %}
+
+By doing so, we ensure that the generated Dockerfile will be compatible with a Bento.
+
+.. dropdown:: About writing :code:`Dockerfile.template`
+   :icon: code
+
+   The Dockerfile template is a mix between :code:`Jinja2` syntax and :code:`Dockerfile`
+   syntax. BentoML set both `trim_blocks` and `lstrip_blocks` in Jinja
+   templates environment to :code:`True`. 
+
+   Make sure that your Dockerfile instruction is **unindented** as if you are writting a normal Dockerfile.
+
+   Refers to `Jinja Whitespace Control <https://jinja.palletsprojects.com/en/3.1.x/templates/#whitespace-control>`_.
+
+   The following example is extracted from BentoML internal templates:
```

Comment:
this example is not that helpful for users, also too complicated -  let's just move the example you currently have in "building bento" to here:

```
{% extends bento__dockerfile %}
{% block SETUP_BENTO_BASE_IMAGE %}
FROM --platform=$BUILDPLATFORM python:3.7-slim as buildstage
RUN mkdir /tmp/mypackage

WORKDIR /tmp/mypackage/
COPY mypackage .
RUN python setup.py sdist && mv dist/mypackage-0.0.1.tar.gz mypackage.tar.gz

{{ super() }}
{% endblock %}
{% block SETUP_BENTO_COMPONENTS %}
{{ super() }}
COPY --from=buildstage mypackage.tar.gz /tmp/wheels/
RUN --network=none pip install --find-links /tmp/wheels mypackage
{% endblock %}
```